### PR TITLE
feature: import implementation for new tracker API

### PR DESCRIPTION
**Pull request by 9sneha-n**
task : https://app.clickup.com/t/860r8gqut
consuming code : https://github.com/EyeSeeTea/glass-dev/blob/ec3687d262f821f31d53b271d58e264d64165bf9/src/data/repositories/TrackerDefaultRepository.ts#L19 
https://github.com/EyeSeeTea/glass-dev/pull/170

### CODE REVIEW

**Review by tokland — COMMENTED** [review]
Looking good, we are very close, just some comments

File: src/api/tracker.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            objectReports: ObjectReports[];
+        };
+    };
+    stats: Stats;
+}
+export interface TrackerPostResponse {
+    status: "OK" | "ERROR" | "WARNING",
+    validationReport: {
+        errorReports: ErrorReport[];
+        warningReports: ErrorReport[];
```
[the remark this answers is not in the document]
Comment (9sneha-n) [reply src/api/tracker.ts:116]: The warning report is exactly the same as Error report as per the DHIS2 code :  https://github.com/dhis2/metadata-package/blob/master/dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerWarningReport.java#L43

File: src/api/tracker.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    };
+    stats: Stats;
+    bundleReport: BundleReport;
+    timingsStats: {},
+    message: string
+}
```
[the remark this answers is not in the document]
Comment (9sneha-n) [reply src/api/tracker.ts:122]: done.

File: src/api/trackerEnrollments.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    getById<Fields extends D2TrackerEnrollmentFields>(
+        params: TrackerEnrollmentParams<Fields>
+    ): D2ApiResponse<GetEnrollment<Fields>> {
+        return this.api.get<GetEnrollment<Fields>>("/tracker/enrollments", {
+            ...params,
+            fields: getFieldsAsString(params.fields),
+        });
+    }
```
[the remark this answers is not in the document]
Comment (9sneha-n) [reply src/api/trackerEnrollments.ts:23]: yes, that makes sense

File: src/api/trackerEnrollments.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    | (Pick<TrackerEnrollmentsParamsBase, "enrolledAfter"> & PartialParams)
+    | (Pick<TrackerEnrollmentsParamsBase, "enrolledBefore"> & PartialParams);
+
+type PartialParams = Partial<TrackerEnrollmentsParamsBase>;
+
+type TrackerEnrollmentsParamsBase = {
+    orgUnit: SemiColonDelimitedListOfUid;
+    ouMode: "SELECTED" | "CHILDREN" | "DESCENDANTS" | "ACCESSIBLE" | "CAPTURE" | "ALL";
+    program: Id;
+    programStatus: "ACTIVE" | "COMPLETED" | "CANCELLED";
```
[the remark this answers is not in the document]
Comment (9sneha-n) [reply src/api/trackerEnrollments.ts:86]: done.

**Review by tokland — APPROVED** [review]
This was ready for review again, right? (you can click on the github action: "re-request review").

All good! Only, on the commit message, I think it's better to do just normal commits (separate by thing and a normal message) instead of a single one referencing the review.

Regarding TrackerEnrollments, I see now, what we had was ok (get all, get by ID), but it was redundant and typically we only provide get (all), which covers the other use case. I've committed some changes and did some tested, worked for me.

So, ready for merging by me :+1:

**Comment by 9sneha-n** [comment]
> 

Hi @tokland , point noted about commit description. It is good for merge from my end too. I am merging the changes, please let me know about releasing a new version.

**Comment by tokland** [comment]
@sneha The release, you are right, there you go: https://www.npmjs.com/package/@eyeseetea/d2-api/v/1.13.2-beta.3

**Comment by 9sneha-n** [comment]
> @sneha The release, you are right, there you go: https://www.npmjs.com/package/@eyeseetea/d2-api/v/1.13.2-beta.3

Thanks @tokland Shall i go ahead and merge the PR then ?

**Comment by adrianq** [comment]
@9sneha-n I am merging it into development instead of master. Note that @tokland release a beta version. Sooner rather than later we will be releasing an official version of d2-api that will contain this and other features. Will give you a shout in one of the slack channels so you can update your app. Thank you for this!

**Pull request merged**